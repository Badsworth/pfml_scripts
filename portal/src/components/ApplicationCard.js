import Claim, { ReasonQualifier } from "../models/Claim";
import Document, { DocumentType } from "../models/Document";
import Alert from "../components/Alert";
import ButtonLink from "../components/ButtonLink";
import Heading from "../components/Heading";
import LeaveReason from "../models/LeaveReason";
import PropTypes from "prop-types";
import React from "react";
import { Trans } from "react-i18next";
import download from "downloadjs";
import findDocumentsByTypes from "../utils/findDocumentsByTypes";
import findKeyByValue from "../utils/findKeyByValue";
import formatDateRange from "../utils/formatDateRange";
import get from "lodash/get";
import hasDocumentsLoadError from "../utils/hasDocumentsLoadError";
import routeWithParams from "../utils/routeWithParams";
import routes from "../routes";
import { useTranslation } from "../locales/i18n";
import withClaimDocuments from "../hoc/withClaimDocuments";

/**
 * Main entry point for an existing benefits Application, allowing
 * claimants to continue an in progress application, view what
 * they've submitted, view notices and instructions, or upload
 * additional docs.
 */
export const ApplicationCard = (props) => {
  const { claim, number } = props;
  const { t } = useTranslation();

  const leaveReason = get(claim, "leave_details.reason");

  return (
    <article className="maxw-mobile-lg border border-base-lighter margin-bottom-3">
      <header className="bg-base-lightest padding-3">
        {claim.fineos_absence_id ? (
          <Heading className="margin-bottom-1" level="2">
            {claim.fineos_absence_id}
          </Heading>
        ) : (
          <Heading className="margin-bottom-05" level="2">
            {t("components.applicationCard.heading", { number })}
          </Heading>
        )}

        {leaveReason && (
          <Heading className="margin-top-0" size="2" level="3" weight="normal">
            {t("components.applicationCard.leaveReasonValue", {
              context: findKeyByValue(LeaveReason, leaveReason),
            })}
          </Heading>
        )}
      </header>

      <div className="padding-3">
        <ApplicationDetails {...props} />
        <LegalNotices {...props} />
        <ApplicationActions {...props} />
      </div>
    </article>
  );
};

ApplicationCard.propTypes = {
  appLogic: PropTypes.shape({
    appErrors: PropTypes.object.isRequired,
    documents: PropTypes.shape({
      download: PropTypes.func.isRequired,
    }),
  }).isRequired,
  claim: PropTypes.instanceOf(Claim).isRequired,
  documents: PropTypes.arrayOf(PropTypes.instanceOf(Document)),
  /**
   * Cards are displayed in a list. What position is this card?
   */
  number: PropTypes.number.isRequired,
};

/**
 * Details about the application, entered by the Claimant,
 * like Employer EIN and leave periods
 */
function ApplicationDetails(props) {
  const { t } = useTranslation();
  const { claim } = props;

  const headingProps = {
    className: "margin-top-0 margin-bottom-05 text-base-dark",
    level: "4",
    size: "6",
  };
  const valueProps = {
    className: "margin-top-0 margin-bottom-2 font-body-2xs text-medium",
  };

  // If an EIN isn't present yet, then this entire component is going to
  // be empty, so we don't want to include a border
  const containerClassName = claim.employer_fein
    ? "border-bottom border-base-lighter margin-bottom-2"
    : undefined;

  return (
    <div className={containerClassName}>
      {claim.employer_fein && (
        <React.Fragment>
          <Heading {...headingProps}>
            {t("components.applicationCard.feinHeading")}
          </Heading>
          <p {...valueProps}>{claim.employer_fein}</p>
        </React.Fragment>
      )}

      {claim.isContinuous && (
        <React.Fragment>
          <Heading {...headingProps}>
            {t("components.applicationCard.leavePeriodLabel_continuous")}
          </Heading>
          <p {...valueProps}>
            {formatDateRange(
              get(
                claim,
                "leave_details.continuous_leave_periods[0].start_date"
              ),
              get(claim, "leave_details.continuous_leave_periods[0].end_date")
            )}
          </p>
        </React.Fragment>
      )}

      {claim.isReducedSchedule && (
        <React.Fragment>
          <Heading {...headingProps}>
            {t("components.applicationCard.leavePeriodLabel_reduced")}
          </Heading>
          <p {...valueProps}>
            {formatDateRange(
              get(
                claim,
                "leave_details.reduced_schedule_leave_periods[0].start_date"
              ),
              get(
                claim,
                "leave_details.reduced_schedule_leave_periods[0].end_date"
              )
            )}
          </p>
        </React.Fragment>
      )}

      {claim.isIntermittent && (
        <React.Fragment>
          <Heading {...headingProps}>
            {t("components.applicationCard.leavePeriodLabel_intermittent")}
          </Heading>
          <p {...valueProps}>
            {formatDateRange(
              get(
                claim,
                "leave_details.intermittent_leave_periods[0].start_date"
              ),
              get(claim, "leave_details.intermittent_leave_periods[0].end_date")
            )}
          </p>
        </React.Fragment>
      )}
    </div>
  );
}

ApplicationDetails.propTypes = {
  claim: PropTypes.instanceOf(Claim).isRequired,
};

/**
 * Legal notices list and content
 */
function LegalNotices(props) {
  const { t } = useTranslation();
  const { appLogic, claim, documents } = props;

  const hasLoadingDocumentsError = hasDocumentsLoadError(
    appLogic.appErrors,
    claim.application_id
  );

  const legalNotices = findDocumentsByTypes(documents, [
    DocumentType.approvalNotice,
    DocumentType.denialNotice,
    DocumentType.requestForInfoNotice,
  ]);

  // If a claim doesn't have a corresponding FINEOS ID, then
  // it's not yet in a state where it could have a legal notice
  return !claim.fineos_absence_id ? null : (
    <div
      className="border-bottom border-base-lighter padding-bottom-2 margin-bottom-2"
      data-test="legal-notices"
    >
      <Heading level="3" weight="normal">
        {t("components.applicationCard.noticesHeading")}
      </Heading>

      {hasLoadingDocumentsError && (
        <Alert noIcon>
          {t("components.applicationCard.documentsRequestError")}
        </Alert>
      )}

      {legalNotices.length === 0 && (
        <p>{t("components.applicationCard.noticesFallback")}</p>
      )}

      {legalNotices.length > 0 && (
        <ul className="usa-list">
          {legalNotices.map((notice) => (
            <LegalNoticeListItem
              key={notice.fineos_document_id}
              document={notice}
              {...props}
            />
          ))}
        </ul>
      )}
    </div>
  );
}

LegalNotices.propTypes = {
  appLogic: PropTypes.shape({
    appErrors: PropTypes.object.isRequired,
  }).isRequired,
  claim: PropTypes.instanceOf(Claim).isRequired,
  documents: PropTypes.arrayOf(PropTypes.instanceOf(Document)),
};

/**
 * Link and metadata for a Legal notice
 */
function LegalNoticeListItem(props) {
  const { appLogic, document } = props;
  const { t } = useTranslation();

  const documentContentType = document.content_type || "application/pdf";
  const noticeNameTranslationKey =
    documentContentType === "application/pdf"
      ? "components.applicationCard.noticeName_pdf"
      : "components.applicationCard.noticeName";

  const handleClick = async (event) => {
    event.preventDefault();
    const documentData = await appLogic.documents.download(document);
    if (documentData) {
      download(
        documentData,
        document.name.trim() || document.document_type.trim(),
        documentContentType
      );
    }
  };

  return (
    <li key={document.fineos_document_id} className="font-body-2xs">
      <a onClick={handleClick} className="text-medium" href="">
        {t(noticeNameTranslationKey, {
          context: findKeyByValue(DocumentType, document.document_type),
        })}
      </a>
      <div className="text-base-dark">
        {t("components.applicationCard.noticeDate", {
          date: formatDateRange(document.created_at),
        })}
      </div>
    </li>
  );
}

LegalNoticeListItem.propTypes = {
  appLogic: PropTypes.shape({
    appErrors: PropTypes.object.isRequired,
    documents: PropTypes.shape({
      download: PropTypes.func.isRequired,
    }),
  }).isRequired,
  document: PropTypes.instanceOf(Document),
};

/**
 * Actions the user can take or will need to take, like calling for
 * reductions, uploading docs, or completing their claim.
 */
function ApplicationActions(props) {
  const { t } = useTranslation();
  const { claim, documents } = props;

  const certificationDocs = findDocumentsByTypes(
    documents,
    // This enum is used because for MVP all certs will have the same doc type
    [DocumentType.medicalCertification]
  );
  const hasDenialNotice =
    findDocumentsByTypes(documents, [DocumentType.denialNotice]).length > 0;

  const hasFutureChildDate = get(claim, "leave_details.has_future_child_date");
  const leaveReasonQualifier = get(claim, "leave_details.reason_qualifier");

  const bondingContentContext = {
    [ReasonQualifier.adoption]: "adopt_foster",
    [ReasonQualifier.fosterCare]: "adopt_foster",
    [ReasonQualifier.newBorn]: "newborn",
  };

  const showBondingLeaveDocRequirement =
    hasFutureChildDate && certificationDocs.length === 0;
  const showResumeButton = !claim.isCompleted && !hasDenialNotice;
  const showUploadButton = claim.isCompleted || hasDenialNotice;

  return (
    <div>
      <Heading level="3" weight="normal">
        {t("components.applicationCard.actionsHeading")}
      </Heading>

      {claim.isCompleted && (
        <React.Fragment>
          <p>
            <Trans
              i18nKey="components.applicationCard.reductionsInstructionsIntro"
              components={{
                "reductions-overview-link": (
                  <a href={routes.external.massgov.reductionsOverview} />
                ),
              }}
            />
          </p>

          <ul className="usa-list">
            <li>{t("components.applicationCard.reductionsInstruction_1")}</li>
            <li>{t("components.applicationCard.reductionsInstruction_2")}</li>
          </ul>
        </React.Fragment>
      )}

      {showBondingLeaveDocRequirement && (
        <p>
          {t("components.applicationCard.bondingLeaveDocsRequired", {
            context: bondingContentContext[leaveReasonQualifier],
          })}
        </p>
      )}

      {showUploadButton && (
        <ButtonLink
          className="display-block"
          href={routeWithParams("applications.uploadDocsOptions", {
            claim_id: claim.application_id,
          })}
          variation={showResumeButton ? "outline" : null}
        >
          {t("components.applicationCard.uploadDocsButton")}
        </ButtonLink>
      )}

      {showResumeButton && (
        <ButtonLink
          data-test="resume-button"
          className="display-block margin-top-2"
          href={routeWithParams("applications.checklist", {
            claim_id: claim.application_id,
          })}
        >
          {t("components.applicationCard.resumeClaimButton")}
        </ButtonLink>
      )}
    </div>
  );
}

ApplicationActions.propTypes = {
  claim: PropTypes.instanceOf(Claim).isRequired,
  documents: PropTypes.arrayOf(PropTypes.instanceOf(Document)),
};

export default withClaimDocuments(ApplicationCard);
