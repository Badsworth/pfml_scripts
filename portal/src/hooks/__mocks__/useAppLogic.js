import Claim from "../../models/Claim";
import ClaimCollection from "../../models/ClaimCollection";
import { uniqueId } from "lodash";

export default jest.fn(() => ({
  auth: {
    createAccount: jest.fn(),
    forgotPassword: jest.fn(),
    login: jest.fn(),
  },
  claims: new ClaimCollection(),
  createClaim: jest.fn(() => new Claim({ application_id: uniqueId() })),
  loadClaims: jest.fn(),
  submitClaim: jest.fn(),
  updateClaim: jest.fn(),
}));
