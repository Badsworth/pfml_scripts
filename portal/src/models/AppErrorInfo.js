import BaseModel from "./BaseModel";
import uniqueId from "lodash/uniqueId";

/**
 * Provides a consistent interface for creating error messages displayed to the user.
 * @property {string} [field] - Path of the field the error is associated with
 * @property {string} key - Unique key used for React `key` prop (https://reactjs.org/docs/lists-and-keys.html)
 * @property {string} message - Internationalized message displayed to the user (like `Error#message`)
 * @property {string} [name] - Name of the error (like `Error#name`)
 */
class AppErrorInfo extends BaseModel {
  get defaults() {
    return {
      key: uniqueId("AppErrorInfo"),
      message: null,
      name: null,
      field: null,
    };
  }
}

export default AppErrorInfo;
