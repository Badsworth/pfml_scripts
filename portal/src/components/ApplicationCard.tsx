import BenefitsApplication, {
  ReasonQualifier,
} from "../models/BenefitsApplication";

import Alert from "./Alert";
import BenefitsApplicationDocument from "../models/BenefitsApplicationDocument";
import ButtonLink from "./ButtonLink";
import { DocumentType } from "../models/Document";
import DownloadableDocument from "./DownloadableDocument";
import Heading from "./Heading";
import LeaveReason from "../models/LeaveReason";
import React from "react";
import { Trans } from "react-i18next";
import findDocumentsByTypes from "../utils/findDocumentsByTypes";
import findKeyByValue from "../utils/findKeyByValue";
import formatDateRange from "../utils/formatDateRange";
import { get } from "lodash";
import hasDocumentsLoadError from "../utils/hasDocumentsLoadError";
import routeWithParams from "../utils/routeWithParams";
import routes from "../routes";
import { useTranslation } from "../locales/i18n";
import withClaimDocuments from "../hoc/withClaimDocuments";

interface ApplicationCardProps {
  appLogic: {
    appErrors: any;
    documents?: {
      download: (...args: any[]) => any;
    };
  };
  claim: BenefitsApplication;
  documents: BenefitsApplicationDocument[];
  /**
   * Cards are displayed in a list. What position is this card?
   */
  number: number;
}

/**
 * Main entry point for an existing benefits Application, allowing
 * claimants to continue an in progress application, view what
 * they've submitted, view notices and instructions, or upload
 * additional docs.
 */
export const ApplicationCard = (props: ApplicationCardProps) => {
  const { claim, number } = props;
  const { t } = useTranslation();

  const leaveReason = get(claim, "leave_details.reason");

  return (
    <article className="maxw-mobile-lg border border-base-lighter margin-bottom-3">
      <header className="bg-base-lightest padding-3">
        {claim.fineos_absence_id ? (
          <Heading className="margin-bottom-1" level="2">
            {claim.fineos_absence_id}
          </Heading>
        ) : (
          <Heading className="margin-bottom-05" level="2">
            {t("components.applicationCard.heading", { number })}
          </Heading>
        )}

        {leaveReason && (
          <Heading className="margin-top-0" size="2" level="3" weight="normal">
            {t("components.applicationCard.leaveReasonValue", {
              context: findKeyByValue(LeaveReason, leaveReason),
            })}
          </Heading>
        )}
      </header>

      <div className="padding-3">
        <ApplicationDetails {...props} />
        <LegalNotices {...props} />
        <ApplicationActions {...props} />
      </div>
    </article>
  );
};

interface ApplicationDetailsProps {
  claim: BenefitsApplication;
}

/**
 * Details about the application, entered by the Claimant,
 * like Employer EIN and leave periods
 */
function ApplicationDetails(props: ApplicationDetailsProps) {
  const { t } = useTranslation();
  const { claim } = props;

  const valueProps = {
    className: "margin-top-0 margin-bottom-2 font-body-2xs text-medium",
  };

  // If an EIN isn't present yet, then this entire component is going to
  // be empty, so we don't want to include a border
  const containerClassName = claim.employer_fein
    ? "border-bottom border-base-lighter margin-bottom-2"
    : undefined;

  return (
    <div className={containerClassName}>
      {claim.employer_fein && (
        <React.Fragment>
          <Heading
            className="margin-top-0 margin-bottom-05 text-base-dark"
            level="4"
            size="6"
          >
            {t("components.applicationCard.feinHeading")}
          </Heading>
          <p {...valueProps}>{claim.employer_fein}</p>
        </React.Fragment>
      )}

      {claim.isContinuous && (
        <React.Fragment>
          <Heading
            className="margin-top-0 margin-bottom-05 text-base-dark"
            level="4"
            size="6"
          >
            {t("components.applicationCard.leavePeriodLabel_continuous")}
          </Heading>
          <p {...valueProps}>
            {formatDateRange(
              get(
                claim,
                "leave_details.continuous_leave_periods[0].start_date"
              ),
              get(claim, "leave_details.continuous_leave_periods[0].end_date")
            )}
          </p>
        </React.Fragment>
      )}

      {claim.isReducedSchedule && (
        <React.Fragment>
          <Heading
            className="margin-top-0 margin-bottom-05 text-base-dark"
            level="4"
            size="6"
          >
            {t("components.applicationCard.leavePeriodLabel_reduced")}
          </Heading>
          <p {...valueProps}>
            {formatDateRange(
              get(
                claim,
                "leave_details.reduced_schedule_leave_periods[0].start_date"
              ),
              get(
                claim,
                "leave_details.reduced_schedule_leave_periods[0].end_date"
              )
            )}
          </p>
        </React.Fragment>
      )}

      {claim.isIntermittent && (
        <React.Fragment>
          <Heading
            className="margin-top-0 margin-bottom-05 text-base-dark"
            level="4"
            size="6"
          >
            {t("components.applicationCard.leavePeriodLabel_intermittent")}
          </Heading>
          <p {...valueProps}>
            {formatDateRange(
              get(
                claim,
                "leave_details.intermittent_leave_periods[0].start_date"
              ),
              get(claim, "leave_details.intermittent_leave_periods[0].end_date")
            )}
          </p>
        </React.Fragment>
      )}
    </div>
  );
}

interface LegalNoticesProps {
  appLogic: any;
  claim: BenefitsApplication;
  documents: BenefitsApplicationDocument[];
}

/**
 * Legal notices list and content
 */
function LegalNotices(props: LegalNoticesProps) {
  const { t } = useTranslation();
  const { appLogic, claim, documents } = props;

  const hasLoadingDocumentsError = hasDocumentsLoadError(
    appLogic.appErrors,
    claim.application_id
  );

  const legalNotices = findDocumentsByTypes(documents, [
    DocumentType.approvalNotice,
    DocumentType.denialNotice,
    DocumentType.requestForInfoNotice,
    DocumentType.withdrawalNotice,
    DocumentType.appealAcknowledgment,
  ]);

  // If a claim doesn't have a corresponding FINEOS ID, then
  // it's not yet in a state where it could have a legal notice
  return !claim.fineos_absence_id ? null : (
    <div
      className="border-bottom border-base-lighter padding-bottom-2 margin-bottom-2"
      data-test="legal-notices"
    >
      <Heading level="3" weight="normal">
        {t("components.applicationCard.noticesHeading")}
      </Heading>

      {hasLoadingDocumentsError && (
        <Alert noIcon>
          <Trans
            i18nKey="components.applicationCard.documentsLoadError"
            components={{
              "contact-center-phone-link": (
                <a href={`tel:${t("shared.contactCenterPhoneNumber")}`} />
              ),
            }}
          />
        </Alert>
      )}

      {legalNotices.length === 0 && (
        <p>{t("components.applicationCard.noticesFallback")}</p>
      )}

      {legalNotices.length > 0 && (
        <React.Fragment>
          <p>{t("components.applicationCard.noticesDownload")}</p>
          <ul className="usa-list">
            {legalNotices.map((notice) => (
              <li key={notice.fineos_document_id}>
                <DownloadableDocument
                  document={notice}
                  showCreatedAt
                  onDownloadClick={appLogic.documents.download}
                />
              </li>
            ))}
          </ul>
        </React.Fragment>
      )}
    </div>
  );
}

interface ApplicationActionsProps {
  claim: BenefitsApplication;
  documents: BenefitsApplicationDocument[];
}

/**
 * Actions the user can take or will need to take, like calling for
 * reductions, uploading docs, or completing their claim.
 */
function ApplicationActions(props: ApplicationActionsProps) {
  const { t } = useTranslation();
  const { claim, documents } = props;

  const certificationDocs = findDocumentsByTypes(
    documents,
    // TODO (CP-2029): Remove reference to `State managed Paid Leave Confirmation` (DocumentType.certification.medicalCertification) type when it's obsolete
    [
      DocumentType.certification.medicalCertification,
      DocumentType.certification[LeaveReason.bonding],
    ]
  );
  const hasDenialNotice =
    findDocumentsByTypes(documents, [DocumentType.denialNotice]).length > 0;

  const hasFutureChildDate = get(claim, "leave_details.has_future_child_date");
  const leaveReasonQualifier = get(claim, "leave_details.reason_qualifier");

  const bondingContentContext = {
    [ReasonQualifier.adoption]: "adopt_foster",
    [ReasonQualifier.fosterCare]: "adopt_foster",
    [ReasonQualifier.newBorn]: "newborn",
  };

  const showBondingLeaveDocRequirement =
    hasFutureChildDate && certificationDocs.length === 0;
  const showResumeButton = !claim.isCompleted && !hasDenialNotice;
  const showUploadButton = claim.isCompleted || hasDenialNotice;

  /**
   * Which Other Leave instructions to display on ApplicationCard
   * TODO (CP-2354) Remove CC guidance for claims with Part 1 submitted without reductions data
   */
  const hasReductionsData =
    get(claim, "has_employer_benefits") !== null ||
    get(claim, "has_other_incomes") !== null ||
    get(claim, "has_previous_leaves_same_reason") !== null ||
    get(claim, "has_previous_leaves_other_reason") !== null ||
    get(claim, "has_concurrent_leave") !== null;
  // Show no instructions by default
  let reductionsI18nKey = null;
  // Show new other leave instructions on completed claims
  if (claim.isCompleted) {
    reductionsI18nKey = "components.applicationCard.reductionsInstructions";
  }
  // Show prompt to report other leaves if Part 1 was submitted with null reductions data
  else if (!hasReductionsData && claim.isSubmitted) {
    reductionsI18nKey =
      "components.applicationCard.reductionsInstructions_missingData";
  }

  return (
    <div>
      <Heading level="3" weight="normal">
        {t("components.applicationCard.actionsHeading")}
      </Heading>

      {reductionsI18nKey && (
        <Trans
          i18nKey={reductionsI18nKey}
          components={{
            "contact-center-phone-link": (
              <a href={`tel:${t("shared.contactCenterPhoneNumber")}`} />
            ),
            "when-can-i-use-pfml": (
              <a href={routes.external.massgov.whenCanIUsePFML} />
            ),
            "reductions-overview-link": (
              <a href={routes.external.massgov.reductionsOverview} />
            ),
            ul: <ul className="usa-list" />,
            li: <li />,
          }}
        />
      )}

      {showBondingLeaveDocRequirement && (
        <p>
          {t("components.applicationCard.bondingLeaveDocsRequired", {
            context: bondingContentContext[leaveReasonQualifier],
          })}
        </p>
      )}

      {showUploadButton && (
        <ButtonLink
          className="display-block"
          href={routeWithParams("applications.uploadDocsOptions", {
            claim_id: claim.application_id,
          })}
          variation={showResumeButton ? "outline" : null}
        >
          {t("components.applicationCard.uploadDocsButton")}
        </ButtonLink>
      )}

      {showResumeButton && (
        <ButtonLink
          data-test="resume-button"
          className="display-block margin-top-2"
          href={routeWithParams("applications.checklist", {
            claim_id: claim.application_id,
          })}
        >
          {t("components.applicationCard.resumeClaimButton")}
        </ButtonLink>
      )}
    </div>
  );
}

export default withClaimDocuments(ApplicationCard);
