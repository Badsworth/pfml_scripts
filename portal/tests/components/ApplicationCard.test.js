import { cleanup, render, screen } from "@testing-library/react";
import AppErrorInfo from "../../src/models/AppErrorInfo";
import { ApplicationCard } from "../../src/components/ApplicationCard";
import { DocumentType } from "../../src/models/Document";
import { MockBenefitsApplicationBuilder } from "../test-utils";
import React from "react";
import User from "../../src/models/User";
import { merge } from "lodash";
import useAppLogic from "../../src/hooks/useAppLogic";

const ApplicationCardWithAppLogic = ({
  // eslint-disable-next-line react/prop-types
  addAppLogicMocks = (appLogic) => {},
  ...otherProps
}) => {
  const appLogic = useAppLogic();
  appLogic.auth.requireLogin = jest.fn();
  appLogic.users.requireUserConsentToDataAgreement = jest.fn();
  appLogic.users.requireUserRole = jest.fn();
  appLogic.users.user = new User({ consented_to_data_sharing: true });

  appLogic.documents.loadAll = jest.fn();
  appLogic.documents.isLoadingClaimDocuments = jest.fn();

  addAppLogicMocks(appLogic);

  const defaultProps = {
    isLoadingDocuments: false,
  };

  const props = merge({}, defaultProps, otherProps);

  return <ApplicationCard appLogic={appLogic} {...props} />;
};

describe("ApplicationCard", () => {
  it("renders In Progress heading and body when application lacks EIN and leave reason", () => {
    const claim = new MockBenefitsApplicationBuilder().create();
    const { container } = render(<ApplicationCardWithAppLogic claim={claim} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("only renders the 'submit all parts' text when application lacks EIN", () => {
    let claim = new MockBenefitsApplicationBuilder().create();
    render(<ApplicationCardWithAppLogic claim={claim} />);

    expect(screen.getByText(/Submit all three parts/i)).toBeInTheDocument();
    cleanup();

    claim = new MockBenefitsApplicationBuilder().employed().create();
    render(<ApplicationCardWithAppLogic claim={claim} />);

    expect(
      screen.queryByText(/Submit all three parts/i)
    ).not.toBeInTheDocument();
  });

  it("renders leave reason as heading when set", () => {
    const claim = new MockBenefitsApplicationBuilder()
      .bondingBirthLeaveReason()
      .create();
    render(<ApplicationCardWithAppLogic claim={claim} />);

    expect(
      screen.getByRole("heading", { level: 2 }).textContent
    ).toMatchInlineSnapshot(`"Leave to bond with a child"`);
  });

  it("renders EIN and absence case id when set", () => {
    const claim = new MockBenefitsApplicationBuilder().submitted().create();

    render(<ApplicationCardWithAppLogic claim={claim} />);

    expect(screen.getByText(claim.employer_fein)).toBeInTheDocument();
    expect(screen.getByText(claim.fineos_absence_id)).toBeInTheDocument();
  });

  it("renders status button and Other Actions when status is completed", () => {
    const claim = new MockBenefitsApplicationBuilder().completed().create();
    const { container } = render(<ApplicationCardWithAppLogic claim={claim} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it.each([
    [
      // "Started"
      new MockBenefitsApplicationBuilder().create(),
      false,
    ],
    [
      // "Submitted"
      new MockBenefitsApplicationBuilder().submitted().create(),
      true,
    ],
    [
      // "Completed"
      new MockBenefitsApplicationBuilder().completed().create(),
      false,
    ],
  ])(
    "only loads documents when status is Submitted",
    (claim, expectDocumentsLoad) => {
      const loadAll = jest.fn();
      render(
        <ApplicationCardWithAppLogic
          addAppLogicMocks={(appLogic) => {
            appLogic.documents.loadAll = loadAll;
            appLogic.documents.isLoadingClaimDocuments = () => true;
          }}
          claim={claim}
          documents={[]}
        />
      );

      if (expectDocumentsLoad) {
        expect(loadAll).toHaveBeenCalledWith(claim.application_id);
        expect(screen.queryByRole("progressbar")).toBeInTheDocument();
      } else {
        expect(loadAll).not.toHaveBeenCalled();
        expect(screen.queryByRole("progressbar")).not.toBeInTheDocument();
      }
    }
  );

  it("does not show a spinner if there is a document load error", () => {
    const claim = new MockBenefitsApplicationBuilder().submitted().create();
    const appErrors = [
      new AppErrorInfo({
        name: "DocumentsLoadError",
        meta: {
          application_id: "mock_application_id",
        },
      }),
    ];

    render(
      <ApplicationCardWithAppLogic
        addAppLogicMocks={(appLogic) => {
          appLogic.appErrors = appErrors;
        }}
        claim={claim}
        documents={[]}
      />
    );

    expect(screen.queryByRole("progressbar")).not.toBeInTheDocument();
  });

  it("doesn't display View notices action if application is Submitted but not Completed", () => {
    const claim = new MockBenefitsApplicationBuilder().submitted().create();
    render(
      <ApplicationCardWithAppLogic
        claim={claim}
        documents={[
          {
            application_id: claim.application_id,
            document_type: DocumentType.appealAcknowledgment,
            fineos_document_id: "mock-document-1",
          },
          {
            application_id: claim.application_id,
            document_type: DocumentType.approvalNotice,
            fineos_document_id: "mock-document-3",
          },
        ]}
      />
    );
    expect(
      screen.queryByRole("button", { name: /View your notices/ })
    ).not.toBeInTheDocument();
  });
});
