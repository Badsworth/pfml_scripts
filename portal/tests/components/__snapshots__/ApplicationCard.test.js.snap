// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationCard renders Alert inside the component when there is an error loading documents 1`] = `"An error was encountered while checking your application for documents. If this continues to happen, call the Paid Family Leave Contact Center at (833) 344‑7365."`;

exports[`ApplicationCard renders EIN 1`] = `
<div
  className="border-bottom border-base-lighter margin-bottom-2"
>
  <Heading
    className="margin-top-0 margin-bottom-05 text-base-dark"
    level="4"
    size="6"
  >
    Employer EIN
  </Heading>
  <p
    className="margin-top-0 margin-bottom-2 font-body-2xs text-medium"
  >
    12-3456789
  </p>
</div>
`;

exports[`ApplicationCard renders Intermittent leave period date range 1`] = `
<div>
  <Heading
    className="margin-top-0 margin-bottom-05 text-base-dark"
    level="4"
    size="6"
  >
    Intermittent leave
  </Heading>
  <p
    className="margin-top-0 margin-bottom-2 font-body-2xs text-medium"
  >
    2/1/2021 – 7/1/2021
  </p>
</div>
`;

exports[`ApplicationCard renders continuous and reduced schedule leave period date ranges 1`] = `
<div>
  <Heading
    className="margin-top-0 margin-bottom-05 text-base-dark"
    level="4"
    size="6"
  >
    Continuous leave
  </Heading>
  <p
    className="margin-top-0 margin-bottom-2 font-body-2xs text-medium"
  >
    1/1/2021 – 6/1/2021
  </p>
  <Heading
    className="margin-top-0 margin-bottom-05 text-base-dark"
    level="4"
    size="6"
  >
    Reduced leave schedule
  </Heading>
  <p
    className="margin-top-0 margin-bottom-2 font-body-2xs text-medium"
  >
    2/1/2021 – 7/1/2021
  </p>
</div>
`;

exports[`ApplicationCard uses generic text as the main heading 1`] = `
<header
  className="bg-base-lightest padding-3"
>
  <Heading
    className="margin-bottom-05"
    level="2"
  >
    Application 2
  </Heading>
</header>
`;

exports[`ApplicationCard when the claim status is Completed includes a button to upload additional documents 1`] = `
<ButtonLink
  className="display-block"
  href="/applications/upload-docs-options?claim_id=mock_application_id"
  variation={null}
>
  Upload additional documents
</ButtonLink>
`;

exports[`ApplicationCard when the claim status is Completed renders instructions about reductions 1`] = `
Array [
  "If you’ll receive ",
  <a
    href="https://www.mass.gov/info-details/paid-family-and-medical-leave-pfml-benefit-reductions"
    key="1"
  >
    other benefits or income
  </a>,
  ", call the Contact Center at ",
  <a
    href="tel:(833) 344-7365"
    key="3"
  >
    (833) 344‑7365
  </a>,
  ". This includes:",
]
`;

exports[`ApplicationCard when the claim status is Completed renders instructions about reductions 2`] = `
<ul
  className="usa-list"
>
  <li>
    Any benefits from your employer that you’ll be using in addition to Massachusetts paid leave (for example, maternity leave, or paid sick time)
  </li>
  <li>
    Income from any other sources during your leave
  </li>
</ul>
`;

exports[`ApplicationCard when the claim status is Submitted includes a link to complete the claim 1`] = `
<ButtonLink
  className="display-block margin-top-2"
  data-test="resume-button"
  href="/applications/checklist?claim_id=mock_application_id"
>
  Continue application
</ButtonLink>
`;

exports[`ApplicationCard when the claim status is Submitted renders legal notices text 1`] = `
<div
  className="border-bottom border-base-lighter padding-bottom-2 margin-bottom-2"
  data-test="legal-notices"
>
  <Heading
    level="3"
    weight="normal"
  >
    Download your notices
  </Heading>
  <p>
    Once we’ve made a decision, you can download the decision notice here. You’ll also get an email notification.
  </p>
</div>
`;

exports[`ApplicationCard when the claim status is Submitted uses the Case ID as the main heading and includes leave reason 1`] = `
<header
  className="bg-base-lightest padding-3"
>
  <Heading
    className="margin-bottom-1"
    level="2"
  >
    NTN-111-ABS-01
  </Heading>
  <Heading
    className="margin-top-0"
    level="3"
    size="2"
    weight="normal"
  >
    Medical leave
  </Heading>
</header>
`;

exports[`ApplicationCard when there are legal notices displays legal notices and explains that notices download to the device 1`] = `
<p>
  When you click the notice link, the file will download to your device.
</p>
`;

exports[`ApplicationCard when there are legal notices displays legal notices and explains that notices download to the device 2`] = `
<li
  className="font-body-2xs"
  key="mock-document-3"
>
  <a
    className="text-medium"
    href=""
    onClick={[Function]}
  >
    Approval notice (PDF)
  </a>
  <div
    className="text-base-dark"
  >
    Posted 1/1/2021
  </div>
</li>
`;

exports[`ApplicationCard when there are legal notices displays legal notices and explains that notices download to the device 3`] = `
<li
  className="font-body-2xs"
  key="mock-document-4"
>
  <a
    className="text-medium"
    href=""
    onClick={[Function]}
  >
    Denial notice
  </a>
  <div
    className="text-base-dark"
  >
    Posted 1/1/2021
  </div>
</li>
`;

exports[`ApplicationCard when there are legal notices displays legal notices and explains that notices download to the device 4`] = `
<li
  className="font-body-2xs"
  key="mock-document-5"
>
  <a
    className="text-medium"
    href=""
    onClick={[Function]}
  >
    Request for more information (PDF)
  </a>
  <div
    className="text-base-dark"
  >
    Posted 1/1/2021
  </div>
</li>
`;

exports[`ApplicationCard when there is a denial notice includes a button to upload additional documents 1`] = `
<ButtonLink
  className="display-block"
  href="/applications/upload-docs-options?claim_id=mock_application_id"
  variation={null}
>
  Upload additional documents
</ButtonLink>
`;
